### ISSUE TITLE
Additions to the Monitoring API

### ISSUE THREAD

**Issue by acchen97** [elastic/logstash#6831]
Next round of per plugin stats for the monitoring API:

- [ ] Beats input
  - [x] connections - current count of Beats connected https://github.com/logstash-plugins/logstash-input-beats/pull/200
  - [x] peak_connections - peak count of Beats connected https://github.com/logstash-plugins/logstash-input-beats/pull/200
  - [ ] beat_types - Beats currently connected and their counts, i.e. Filebeat 20, Metricbeat 6, etc.
- [x] Dissect filter - matches/failures https://github.com/logstash-plugins/logstash-filter-dissect/issues/13 PR https://github.com/logstash-plugins/logstash-filter-dissect/pull/14
- [x] Date filter - matches/failures https://github.com/logstash-plugins/logstash-filter-date/issues/83 PR: https://github.com/logstash-plugins/logstash-filter-date/pull/92
- [ ] Elasticsearch output
  - [ ] bulk_requests - number of bulk requests sent PR https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585
    - [ ] Success count
    - [ ] Failure count
  - [ ] concurrent_bulk_requests - https://github.com/elastic/logstash/issues/6544 PR https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585
  - [ ] retry_count - number of retries PR https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585
  - [ ] sniffing_count - number of times sniffing was triggered PR https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585
  - [ ] HTTP response code count, buckets i.e. 2xx, 3xx, 4xx, 5xx

/cc @suyograo

**Comment by suyograo** [comment]
I removed the PQ stats from here which we can track separately.

**Comment by acchen97** [comment]
Great, I've opened up a separate issue for the disk stats https://github.com/elastic/logstash/issues/6884.

**Comment by jsvd** [comment]
Output plugin metrics are blocked by https://github.com/elastic/logstash/issues/6889

**Comment by jsvd** [comment]
Just as a heads up: currently logstash core doesn't expose the metrics subsystem in Java, as everything is still in ruby except the LogStash Event class.
This means that, for a plugin written in Java to produce metrics, it must either:
1. use dynamic method lookups to access the ruby metric object in the Java code (done in the dissect filter)
2. use jruby's [proc-to-interface conversion](https://github.com/jruby/jruby/wiki/CallingJavaFromJRuby#closure-conversion) to pass ruby blocks into the java class that the Java code calls in certain parts of its flow (done in the date filter)

This means that producing metrics in java based plugins is a more involved task and may take longer than initially expected.

**Comment by acchen97** [comment]
> This means that, for a plugin written in Java to produce metrics, it must either:
> 
> 1. use dynamic method lookups to access the ruby metric object in the Java code (done in the dissect filter)
> 2. use jruby's proc-to-interface conversion to pass ruby blocks into the java class that the Java code calls in certain parts of its flow (done in the date filter)

@jsvd does it make sense to instead have a single standard way of producing these metrics? I'd guess explicit lookup methods would be preferred as we've been moving towards that direction, like with the Java event.

**Comment by jsvd** [comment]
@acchen97 definitely we need a standard way of doing it, but since java plugins are still a mix of ruby and jruby, depending on the amount of ruby left, the strategy will change. For now we'll have to tackle each one individually, I've already added metrics to date and dissect.

Few questions for elasticsearch output:
1.  request_latency_in_millis 
  a) we already have the commulative time spend doing the filter operation, do we want just the time for the request itself (also commulative?)?
2.  "HTTP response code count, buckets i.e. 2xx, 3xx, 4xx, 5xx"
  a) do we want just the http codes for the individual items of the bulk response, or do we want to track the http code of the bulk response itself? if so, we probably need two different metrics
  b) do we want to do bucketing on the logstash level? since the list of possible codes isn't very big I think we should track them all individually. thoughts?
3.  bulk_requests - we can have different semantics for this metric, I suggest number a)
a) number of bulk requests measured right before executing the request
b) number of bulk requests sent
c) number of bulk requests finished

**Comment by acchen97** [comment]
> request_latency_in_millis
> a) we already have the commulative time spend doing the filter operation, do we want just the time for the request itself (also commulative?)?

When you say "request itself" do you mean just request to response from ES? I would say, if possible, for us to track from the time it enters the output stage to when it's formally acked.

> "HTTP response code count, buckets i.e. 2xx, 3xx, 4xx, 5xx"
> a) do we want just the http codes for the individual items of the bulk response, or do we want to track the http code of the bulk response itself? if so, we probably need two different metrics

Both probably make sense, what do you think @ph? I wasn't aware that individual events had their own http codes.

> b) do we want to do bucketing on the logstash level? since the list of possible codes isn't very big I think we should track them all individually. thoughts?

Yes, that sounds like the right way to go.

> bulk_requests - we can have different semantics for this metric, I suggest number a)
> a) number of bulk requests measured right before executing the request
> b) number of bulk requests sent
> c) number of bulk requests finished

@jsvd I'm not sure what you mean with (a). Isn't it the same as (b), just different time of calculation? I was thinking this could perhaps track the number of successful bulk_requests.

**Comment by jsvd** [comment]
>When you say "request itself" do you mean just request to response from ES? I would say, if possible, for us to track from the time it enters the output stage to when it's formally acked.

This is already tracked by default with `duration_in_millis`, that is why I thought you wanted to track just the time taken to perform the bulk request.

> @jsvd I'm not sure what you mean with (a). Isn't it the same as (b), just different time of calculation? I was thinking this could perhaps track the number of successful bulk_requests.

Let's reduce to two cases, a and b, if you picture the pseude code being:

```ruby
def perform_bulk(data)
  # we can measure the count before the request
  response = @client.bulk(data)
  # or after a success
  return response.body
rescue StandardError => e
   # we can also count the failures
   retry if response.code.retryable?
end
```

We can measure the count before performing the request or after. If we measure after, we need to decide if we're measuring only the successes or successes+failures. In https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585/files#diff-8eecbfb9f47550237e76ea104f7156b5R139 the code measures the count before performing the bulk request, this way we know both the failures and successes without measuring twice after the fact.

**Comment by acchen97** [comment]
> This is already tracked by default with duration_in_millis, that is why I thought you wanted to track just the time taken to perform the bulk request.

Got it, let's remove this metric (I'll remove it from the original issue). The duration_in_millis should cover it.

> We can measure the count before performing the request or after. If we measure after, we need to decide if we're measuring only the successes or successes+failures. In https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/585/files#diff-8eecbfb9f47550237e76ea104f7156b5R139 the code measures the count before performing the bulk request, this way we know both the failures and successes without measuring twice after the fact.

Let's do it after the request. As you lay it out in this context, I think its valuable for us to track both successes and failures separately. Will update the issue on that as well.

**Comment by jsvd** [comment]
@acchen97 going a bit deeper into es output metrics, there are few more possible outcomes of a bulk request that will require tracking. A bulk request can have many outcomes and reactions:

1. host unreachable/no connection available/no HTTP response: request is retried
2. received an non {200,201} HTTP response to the bulk request
2.1. retriable response code: the whole bulk request is retried
2.2. non retriable response code: entire bulk request is dropped
3. received a {200,201} response to bulk request:
3.1. bulk response contains no `response["errors"]`: successful bulk request
3.2. bulk response contains `response["errors"]`: the portion of the bulk with the errors is retried
3.2.1 document-level response error code isn't retriable: document dropped from the bulk

Since there are document level and bulk level notions of success and failures, I propose the following metrics:

```
{
  "bulk_requests": {
     "responses": {
       200: 346300,
       201: 712,
       429: 36
     },
     "successes": 346819,  # successful bulk requests with no errors (sum of 200 and 201 minus partial_successes)
     "failures": 23,         # no http response from ES. will be retried. see point 1 above for causes
     "partial_successes": 134,  # bulk request successful (status 20X) but contains errors
     "drops": 5,   # bulk request dropped
     "document_drops": 12 # documents permanently dropped from bulk requests
  },
  "sniffing_count": 621
}
```

The only thing missing here is the count of HTTP response codes for each individual document in the bulk.
I'm still not super happy with this hierarchy, thoughts?

**Comment by ph** [comment]
@jsvd  I think its a good first try, should we record all the response code like 404 or they will be part of `document_drops`

**Comment by jsvd** [comment]
@ph you mean the response code from individual items in the bulk response? If you mean the top level bulk response code, then yes, we should record all, my example only shows 3, but I'll edit it to explain better.

**Comment by ph** [comment]
@jsvd I was talking about the *bulk_response*, thanks for clarifying :)

**Comment by acchen97** [comment]
@jsvd this is good stuff, thanks!

I wonder if we have to be as granular for the document level stats. Perhaps we move the "document_drops" to a high level documents object, e.g.:

```ruby
{
  ...
  "documents": {
    "successes": 173409500,
    "retries": 450,
    "drops": 12
  }
}
```

As we should have consistent terminology between bulk and documents, should we count "failures" or "retries"? My understanding is a failure will either be retried or dropped. Additionally, what's the behavior when we retry the other docs in a "bulk.requests.partial_successes"? If that succeeds, will it then also increment the "bulk_requests.successes"?

**Comment by jsvd** [comment]
> I wonder if we have to be as granular for the document level stats. Perhaps we move the "document_drops" to a high level documents object, e.g.:

+1 on not being going so deep wrt documents, we can start small(er) and then add them if requested. as for the nested document, we could have:

```
  "documents": {
    "successes": 173409500,
    "retryable_failures": 450,
    "non_retryable_failures": 12
  }
```

> should we count "failures" or "retries"?
The good thing about failures is we can measure them as result of a bulk request. Retries should be measured after the bulk requests fails AND the second bulk request is executed, which makes things more complicated.

> Additionally, what's the behavior when we retry the other docs in a "bulk.requests.partial_successes"? If that succeeds, will it then also increment the "bulk_requests.successes"?

It is up for debate, I think it should. as it is an actual post to _bulk, so a bulk request. the retry mechanism is a construct done by the client (logstash).

**Comment by suyograo** [comment]
Pushing to 5.4.0

**Comment by acchen97** [comment]
@jsvd looks good to me. So the cumulative object would look like this:

```ruby
{
  "bulk_requests": {
     "responses": {  # counts per top level status code returned in the bulk response
       200: 346300,
       201: 712,
       404: 61,
       429: 36
       # other status codes ...
     },
     "successes": 346819,  # successful bulk requests with no errors (sum of 200 and 201 minus partial_successes)
     "failures": 23,         # no http response from ES. will be retried. see point 1 above for causes
     "partial_successes": 134,  # bulk request successful (status 20X) but contains errors
     "drops": 5,   # bulk request dropped
  },
  "documents": {
       "successes": 173409500,
       "retryable_failures": 450,
       "non_retryable_failures": 12
  },
  "sniffing_count": 621
}
```


> It is up for debate, I think it should. as it is an actual post to _bulk, so a bulk request. the retry mechanism is a construct done by the client (logstash).

Agreed, let's do that. I think the only item we still need closure on is the concurrent bulk request metric.

**Comment by jordansissel** [comment]
@acchen97 can you move all the things that didn't make it into 5.5 into a new ticket and leave the ones which landed in 5.5 in this one?

**Comment by acchen97** [comment]
@jsvd do you mind marking the items in the initial issue that were completed? I'm not sure whether we were able to get all the ES output ones in.

**Comment by jsvd** [comment]
@acchen97 the ones marked completed have been published, the missing ones are the ES output (there's a PR open but incomplete) and in the beats input we don't yet have the metrics per beat type

**Comment by acchen97** [comment]
Superceded, with outstanding items moved to https://github.com/elastic/logstash/issues/7621.
